packages/contracts/contracts/plugin/PluginSetupProcessor.sol

```diff
@@ -186,394 +246,388 @@ contract PluginSetupProcessor is DaoAuthorizable {
-    /// @param _currentHelpers The address array of all current helpers (contracts or EOAs) associated with the plugin that is prepared to be uninstalled.
-    /// @param _data The `bytes` encoded data containing the input parameters for the uninstallation as specified in the `prepareUninstallationDataABI()` function in the `pluginSetup` setup contract.
+    /// @param _params The struct containing the parameters for the `prepareUninstallation` function.
     /// @return permissions The list of multi-targeted permission operations to be applied to the uninstalling DAO.
-    /// @dev The list of `_currentHelpers` has to be specified in the same order as they were returned from previous setups preparation steps (the latest `prepareInstallation` or `prepareUpdate` step that has happend) on which the uninstallation was prepared for.
-    function prepareUninstallation(
-        address _dao,
-        address _plugin,
-        address _pluginSetup,
-        PluginRepo _pluginSetupRepo,
-        address[] calldata _currentHelpers,
-        bytes calldata _data
-    ) external returns (PermissionLib.ItemMultiTarget[] memory permissions) {
-        // ensure repo for plugin manager exists
-        if (!repoRegistry.entries(address(_pluginSetupRepo))) {
-            revert PluginRepoNonexistent();
-        }
+    /// @dev The list of `_currentHelpers` has to be specified in the same order as they were returned from previous setups preparation steps (the latest `prepareInstallation` or `prepareUpdate` step that has happend) on which the uninstallation was prepared for
+    function prepareUninstallation(address _dao, PrepareUninstallationParams calldata _params)
+        external
+        returns (PermissionLib.ItemMultiTarget[] memory permissions)
+    {
+        bytes32 pluginId = _getPluginId(_dao, _params.setupPayload.plugin);
 
-        // Reverts if pluginSetup doesn't exist on the repo...
-        _pluginSetupRepo.getVersionByPluginSetup(_pluginSetup);
+        PluginState storage pluginState = states[pluginId];
 
-        // check if plugin is applied.
-        bytes32 appliedId = _getAppliedId(_dao, _plugin);
+        bytes32 setupId = _getSetupId(
+            _params.pluginSetupRef,
+            bytes32(0),
+            hashHelpers(_params.setupPayload.currentHelpers),
+            bytes(""),
+            PreparationType.None
+        );
 
-        if (!isInstallationApplied[appliedId]) {
-            revert SetupNotApplied();
+        if (pluginState.currentSetupId != setupId) {
+            revert InvalidSetupId({currentSetupId: pluginState.currentSetupId, setupId: setupId});
         }
 
-        permissions = PluginSetup(_pluginSetup).prepareUninstallation(
-            _dao,
-            _plugin,
-            _currentHelpers,
-            _data
+        PluginRepo.Version memory version = _params.pluginSetupRef.pluginSetupRepo.getVersion(
```

Comment:
why we don't have a local variable similar `prepareInstallation()` like : `PluginRepo pluginSetupRepo = _params.pluginSetupRef.pluginSetupRepo;`

Reply:
trying my best not to store stuff on stack. Resolve if you think the answer is acceptable.

Reply:
Won't the compiler end up using the stack anyway?

Local variables or complex expressions, they generally end up using the stack or a register (depending on the architecture).

```
let a = 1;
let b = 2;
let c = 3;
let d = 4; 

return (a-c) ^ ((b+d+a) + (c-a-b) / (a*b*d)) ;
```

And this:

```
let a = 1, b = 2, c = 3, d = 4;

const a_c = a-c;
const b_d_a = b+d+a;
const c_a_b = c-a-b;
const a_b_d = a*b*d;

return a_c ^ (b_d_a + c_a_b / a_b_d) ;
```

Should produce the same instructions in most cases.

Adding my 5c, whatever you feel is best

Reply:
If I do what sarkawt is saying, the variable will stay on stack until the function is finished(This really works untuitively, sometimes optimizer can figure it out that the variable is not used anymore in the rest remaining function, but sometimes it can't). 

What I do though, is it will be put on stack, but as soon as `getVersion` call is done, it will remove it. NOTE that we're talking about storing it in `address pluginSetupRepo`;

Reply:
the aim was just to bring ur attention to it, since it assigned to local variable in `prepareInstallation()`
if u think it is fine, feel free to resolve this comment